### ISSUE TITLE
decorator false positive: unsubscriptable-object, no-member, unsupported-membership-test, not-an-iterable, too-many-function-args

### ISSUE THREAD

**Issue by left-join** [PyCQA/pylint#1694]
### Steps to reproduce
I use simple classproperty decorator:
```python
class classproperty(classmethod):

    def __init__(self, fget):
        if isinstance(fget, (classmethod, staticmethod)):
            self.fget = lambda cls: fget.__get__(None, cls)()
        else:
            self.fget = fget
        self.cached = {}
        super(classproperty, self).__init__(self.fget)

    def __get__(self, instance, cls):
        if cls in self.cached:
            return self.cached[cls]
        value = self.cached[cls] = self.fget(cls)
        return value

```

Example of usages:
```python
from collections import OrderedDict

from sqlalchemy.ext.declarative import declarative_base, declared_attr
from sqlalchemy.inspection import inspect

Base = declarative_base()

class Model(Base):
    __abstract__ = True

    def __init__(self, **kwargs):
        super(Model, self).__init__()
        self.set_attrs(**kwargs)

    def set_attrs(self, **attrs):
        for attr_name, attr_value in attrs.items():
            if attr_name in self.fields:
                setattr(self, attr_name, attr_value)

    def to_dict(self):
        result = {}
        for attr_name in self.fields:
            attr_value = getattr(self, attr_name)
            if attr_value is not None:
                result[attr_name] = attr_value
        return result

    @declared_attr
    def __tablename__(cls):
        # pylint: disable=no-self-argument
        return plural(decapitalize(cls.__name__))

    @classproperty
    def columns(cls):
        columns = inspect(cls).mapper.column_attrs
        columns = list(sorted(columns, key=lambda column: column.key))
        return columns

    @classproperty
    def fields(cls):
        fields = OrderedDict([(column.key, getattr(cls, column.key))
                              for column in cls.columns])
        return fields

    @classproperty
    def primary_key_columns(cls):
        columns = list(inspect(cls).mapper.primary_key)
        return columns

    @classproperty
    def primary_key_fields(cls):
        fields = OrderedDict([(column.key, getattr(cls, column.key))
                              for column in cls.primary_key_columns])
        return fields

```

### Current behavior
There are a lot of false positive errors:

```bash
[E1135(unsupported-membership-test), Model.set_attrs] Value 'self.fields' doesn't support membership test
[E1133(not-an-iterable), Model.to_dict] Non-iterable value self.fields is used in an iterating context
[E1133(not-an-iterable), Model.fields] Non-iterable value cls.columns is used in an iterating context
[E1133(not-an-iterable), Model.primary_key_fields] Non-iterable value cls.primary_key_columns is used in an iterating context
````

### Expected behavior
If I make interface class where declare expected value the errors are not raised:
```python

class iface(object):
    columns = ()
    fields = {}
    primary_key_columns = ()
    primary_key_fields = {}

```
Expected behavior: no false-positive errors

### pylint --version output
pylint 1.7.4, 
astroid 1.5.3
Python 3.5.2 (default, Sep 14 2017, 22:51:06) 
[GCC 5.4.0 20160609]

**Comment by PCManticore** [comment]
This is a known issue, we don't yet support custom decorators/descriptors.

**Comment by haxwithaxe** [comment]
I'd like to write this but I'm not finding any unit/functional tests. Am I missing something?

**Comment by jacobtylerwalls** [comment]
Test case from #818 involving the Voluptuous library:
```python
"""Demonstration of no-value-for-parameter using Voluptuous"""
#
# This code shows the proper use of the python Voluptuous
# library.  IsFile() is a callable that can be passed
# as a dictionary value in a schema definition.  When called
# it validates the parameter that is passed in.
#
# It is defined as
#    def IsFile(v):
# but called from within Voluptuous as
#    IsFile()('passed param')
#
# Pylint finds
#
# E: 24,39: No value for argument 'v' in function call (no-value-for-parameter)
#
from voluptuous import Schema, IsFile, MultipleInvalid

def main():
    """Testing Voluptuous IsFile() validator"""

    data = {'filename': 'not a file.txt'}

    schema = Schema({'filename': IsFile()})

    try:
        schema(data)
    except MultipleInvalid as e:
        print("Success:", e.msg)
    else:
        print("Failure: IsFile did not work")

if __name__ == "__main__":
    main()
```
